compose/service.py

```diff
@@ -1138,6 +1138,12 @@ def get_cache_from(self, build_opts):
     def can_be_built(self):
         return 'build' in self.options
 
+    def can_be_pulled(self):
+        return 'image' in self.options
+
+    def can_be_pushed(self):
+        return 'build' in self.options and 'image' in self.options
```

Comment:
Not sure if there's a validation for this already, but `image` can also have a _digest_ in it (e.g. `busybox@sha256:a8cf7ff6367c2afa2a90acd081b484cbded349a7076e7bdf37a05279f276bc12`)

While that's valid for _pulling_ an image, it's not valid for _building_ or _pushing_ an image; I think this function should check if `parse_repository_tag ` returns `digest_separator` (`@`) as separator, and if so, return `false`.

Probably something like this.

```python
def can_be_pushed(self):
    if 'build' not in self.options or 'image' not in self.options:
        return False
    _, _, separator = parse_repository_tag(self.options.get('image'))
    return separator != '@'
```

Can you also add a test for these?

Reply:
On a side-note: from my limited Python skills it looks like currently, `parse_repository_tag` does not support a reference to have _both_ a `tag` and a `digest`, for example:

```
busybox:latest@sha256:a8cf7ff6367c2afa2a90acd081b484cbded349a7076e7bdf37a05279f276bc12 
```

admittedly, that syntax is mostly is to make the reference "nicer" to read, because _digest_ is leading in that case, but it probably should be looked at 
https://github.com/docker/compose/blob/440c94ea7a7e62b3de50722120ca34c4e818205a/compose/service.py#L1443-L1445

/cc @ulyssessouza

Reply:
> Not sure if there's a validation for this already, but `image` can also have a _digest_ in it (e.g. `busybox@sha256:a8cf7ff6367c2afa2a90acd081b484cbded349a7076e7bdf37a05279f276bc12`)
> 
> While that's valid for _pulling_ an image, it's not valid for _building_ or _pushing_ an image; I think this function should check if `parse_repository_tag ` returns `digest_separator` (`@`) as separator, and if so, return `false`.
> 
> Probably something like this.
> 
> ```python
> def can_be_pushed(self):
>     if 'build' not in self.options or 'image' not in self.options:
>         return False
>     _, _, separator = parse_repository_tag(self.options.get('image'))
>     return separator != '@'
> ```
> 
> Can you also add a test for these?

Well, if someone stated, that service has `build` phase and `image` with digest, it's a mistake in config. `push` handler shouldn't be bothered by it, the `client` will fail on push with explanation.
The better way is to add such checks in config validation IMHO

Reply:
(update) suggestion to leave can_be_pushed same, but add exception if trying to push image with digest
```
    def can_be_pushed(self):
        return 'build' in self.options and 'image' in self.options

...

    def push(self, ignore_push_failures=False):
        if not self.can_be_pushed():
            return

        repo, tag, separator = parse_repository_tag(self.options['image'])
        if separator == '@':
            raise ConfigurationError('Can\'t push image with digest')
        tag = tag or 'latest'
        log.info('Pushing %s (%s%s%s)...' % (self.name, repo, separator, tag))
        output = self.client.push(repo, tag=tag, stream=True)
        ...
```

Reply:
> On a side-note: from my limited Python skills it looks like currently, `parse_repository_tag` does not support a reference to have _both_ a `tag` and a `digest`, for example:
> 
> ```
> busybox:latest@sha256:a8cf7ff6367c2afa2a90acd081b484cbded349a7076e7bdf37a05279f276bc12 
> ```
> 
> admittedly, that syntax is mostly is to make the reference "nicer" to read, because _digest_ is leading in that case, but it probably should be looked at
> https://github.com/docker/compose/blob/440c94ea7a7e62b3de50722120ca34c4e818205a/compose/service.py#L1443-L1445
> 
> /cc @ulyssessouza

What is excepted behaviour? Jsut ignore tag, so
```python
>>> parse_repository_tag('user/repo:v1@sha256:digest')
('user/repo', 'sha256:digest', '@')
```

Reply:
I think pulling with just the digest is ok (digest references can only be used for pulling / running, not for pushing/building). Unless we want this function to return all things separate (which may be more "correct", but would require more changes);

This is "ok" `image@sha256:deadbeef`

But more "correct" would be 


```
repo, tag, digest = parse_repository_tag('user/repo:v1@sha256:digest')
```

(return the tag (if any) and/or digest (if any)); perhaps it should return some struct having those properties.

But (as mentioned) for now, just using `image@sha256:deadbeef` should be ok

Reply:
> But (as mentioned) for now, just using `image@sha256:deadbeef` should be ok

Added #7432 pull request for it, 'cause it's separate functionality